corehq/apps/es/README.rst

```diff
@@ -0,0 +1,220 @@
+Elasticsearch App
+=================
+
+Adapter Design
+--------------
+
+The HQ Elastic adapter design came about due to the need for reindexing
+Elasticsearch indexes in a way that is transparent to parts of HQ that write to
+Elasticsearch (e.g. pillowtop).  Reindexing is necessary for making changes to
+index mappings, is a prerequisite to upgrading an Elasticsearch cluster, and is
+also needed for changing low-level index configurations (e.g. sharding).
+
+There is an existing procedure draft that documents the steps that were used on
+one occasion to reindex the ``case_search`` index.  This procedure leveraged a
+custom pillow to "backfill" the cloned index (i.e. initially populated using
+Elasticsearch Reindex API). That procedure only works for a subset of HQ
+Elasticsearch indexes, and is too risky to be considered as an ongoing Elastic
+maintenance strategy. There are several key constraints that an HQ reindexing
+procedure should meet which the existing procedure does not:
+
+- simple and robust
+- performed with standard maintenance practices
+- provides the ability to test and verify the integrity of a new index before it
+  is too late to be rejected
+- allows HQ Elasticsearch index state to remain decoupled from the
+  commcare-cloud codebase
+- is not disruptive -- does not prohibit any other kind of standard maintenance
+  that might come up while the operation is underway
+- is "fire and forget" -- does not require active polling of intermediate state
+  in order to progress the overall operation
+- is practical for third party HQ hosters to use
+
+One way to accomplish these constraints is to implement an "index multiplexing"
+feature in HQ, where Elasticsearch write operations are duplicated across two
+indexes. This design facilitates maintaining two up-to-date versions of any
+index (a primary read/write index and a secondary write-only index),
+allowing HQ to run in a "normal" state (i.e. not a custom "maintenance" state)
+while providing the ability to switch back and forth (swapping primary and
+secondary) before fully committing to abandoning one of them. Creating a copy of
+an index is the unavoidable nature of a reindex operation, and multiplexing
+allows safe switching from one to the other without causing disruptions or
+outages while keeping both up-to-date.
```

Comment:
> In the past we've accomplished this with temporary pillows to 'top up' the new index and then again to top up the old index after the switchover.

We have not used this process on all indexes. As far as I can tell we've only used it on one: the `case_search` index, which has a much simpler pillow configuration and usage than say, `users`, or `forms`.  This process simply will not work for all indexes, and is why a new process is needed.  This old process also has no mechanism for verifying data integrity of the index after the "top up" is complete.

Reply:
> This process simply will not work for all indexes

I don't see why not. Why is the complexity of the pillow an issue? It's simple enough to create temporary pillows that only use specific processors. In fact we still have most of the original single processor pillow configurations in code.

> This old process also has no mechanism for verifying data integrity of the index after the "top up" is complete.

How do you intend to do that with the new process?

Reply:
> Why is the complexity of the pillow an issue?

It's error-prone and hard for an operator to do properly.

Reply:
> How do you intend to do that with the new process?

By spot-checking documents between the old (primary) and new (secondary) indexes.

Reply:
To me this feels like a critical piece, we've tried this before in some ways and without a lot of success.

There is plenty of theory on population sampling so it would be great to figure out some of the details here so that we can be confident it's going to work.